docs/design/datacontracts/contract_csharp_api_design.cs

```diff
@@ -0,0 +1,386 @@
+        bool TryReadGlobalUInt32(string globalName, out uint value);
+        bool TryReadGlobalInt64(string globalName, out long value);
+        bool TryReadGlobalUInt64(string globalName, out ulong value);
+        bool TryReadGlobalTargetPointer(string globalName, out TargetPointer value);
+
+        Contracts Contract { get; }
+
+        partial class Contracts
+        {
+            FieldLayout GetFieldLayout(string typeName, string fieldName);
+            bool TryGetFieldLayout(string typeName, string fieldName, out FieldLayout layout);
+            int GetTypeSize(string typeName);
+            bool TryGetTypeSize(string typeName, out int size);
+
+            object GetContract(string contractName);
+            bool TryGetContract(string contractName, out object contract);
+
+            // Every contract that is defined has a field here. As an example this document defines a MethodTableContract
+            // If the contract is not supported by the runtime in use, then the implementation of the contract will be the base type which
+            // is defined to throw if it is ever used.
+
+            // List of contracts will be inserted here by source generator
+            MethodTableContract MethodTableContract;
+        }
+    }
+
+    // Types defined by contracts live here
+    namespace ContractDefinitions
+    {
+        class CompositeContract
+        {
+            List<Tuple<string, uint>> Subcontracts;
+        }
+
+        class DataStructureContract
+        {
+            string MethodTableName {get;}
+            List<Tuple<string, FieldLayout>> FieldData;
+        }
+
+        // Insert Algorithmic Contract definitions here
+        class MethodTableContract
+        {
+            public virtual int DynamicTypeID(TargetPointer methodTablePointer) { throw new NotImplementedException(); }
+            public virtual int BaseSize(TargetPointer methodTablePointer) { throw new NotImplementedException(); }
+        }
+    }
+
+    namespace ContractImplementation
+    {
+        // Get contract from the predefined contract database
+        static class PredefinedContracts
+        {
+            public static IContract GetContract(string name, uint version, Target target)
+            {
+                // Do some lookup and allocate an instance of the contract requested
+                //
+                // This lookup can either be reflection based, or we can do it based on a source generator.
+            }
+        }
+
+        [DataContractGlobals("FeatureFlags", 1)]
+        public class FeatureFlags_1
+        {
+            public const int FeatureComInterop = 0;
+        }
+
+        [DataContractGlobals("FeatureFlags", 2)]
+        public class FeatureFlags_2
+        {
+            public const int FeatureComInterop = 1;
+        }
+
+        [DataContractAlgorithm(1)]
+        class MethodTableContract_1 : ContractDefinitions.MethodTableContract, IAlgorithmContract
```

Comment:
I'm not sure why some contract info is here and other contract info is in markdown files. It seems like there are two different patterns in use. Given a choice between the two I think the markdown approach is easier to read and understand.

Reply:
Hmm... I think this is a bit misunderstood... the concept of a "MehtodTableContract_1" was intended as an example of what the final code would be. I really ought to put a comment here.

Reply:
Yeah, all the other content here appeared to be examples of how we'd document contracts so that was my default interpretation of the content here too. Since this is intended to be something else we could just put some comments at the top of the file saying so.